DRAFT: Optimized Non-Conservative Zonal Average

hongyuchen1030 opened the pull request:
<!--  The PR title should summarize the changes, for example "Add `Grid._build_face_dimension` function".
      Avoid non-descriptive titles such as "Addresses issue #229". -->

<!--  Replace XXX with the number of the issue that this PR will resolve. If this PR closed more than one,
      you may add a comma separated sequence-->
Closes #1181 #880 

## Overview
- Introduce specialized helper functions for the optimized non-conservative zonal average routine in `zonal.py`
- Implement caching and reuse of intermediary computational steps to reduce redundant recalculations
- Retain the general-purpose helper functions (which use data copying) for broader usability
- Improve overall performance of the zonal averaging computation


## Expected Usage
<!--  If you are adding a feature into the Internal API, please produce a short example of it in action.
      You may ignore this step if it is not applicable (comment out this section). -->
```Python
import uxarray as ux

grid_path = "/path/to/grid.nc"
data_path = "/path/to/data.nc"

uxds = ux.open_dataset(grid_path, data_path)

# this is how you use this function
some_output = uxds.some_function()

# this is another way to use this function
other_output = uxds.some_function(some_param = True)

```

## PR Checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. If an entire section doesn't
apply to this PR, comment it out or delete it. -->

**General**
- [x] An issue is linked created and linked
- [x] Add appropriate labels
- [ ] Filled out Overview and Expected Usage (if applicable) sections

**Testing**
- [ ] Adequate tests are created if there is new functionality
- [ ] Tests cover all possible logical paths in your function
- [ ] Tests are not too basic (such as simply calling a function and nothing else)

**Documentation**
- [ ] Docstrings have been added to all new functions
- [ ] Docstrings have updated with any function changes
- [ ] Internal functions have a preceding underscore (`_`) and have been added to `docs/internal_api/index.rst`
- [ ] User functions have been added to `docs/user_api/index.rst`

**Examples**
- [ ] Any new notebook examples added to `docs/examples/` folder
- [ ] Clear the output of all cells before committing
- [ ] New notebook files added to `docs/examples.rst` toctree
- [ ] New notebook files added to new entry in `docs/gallery.yml` with appropriate thumbnail photo in `docs/_static/thumbnails/`

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

**PR Etiquette Reminders**
- This PR should be listed as a draft PR until you are ready to request reviewers

- After making changes in accordance with the reviews, re-request your reviewers

- Do *not* mark conversations as resolved if you didn't start them

- Do mark conversations as resolved *if you opened them* and are satisfied with the changes/discussion.
-->

hongyuchen1030 commented on uxarray/grid/geometry.py:
```diff
@@ -1407,25 +1466,14 @@ def _populate_bounds(
     """
 
     # Ensure grid's cartesian coordinates are normalized
+    # TODO: Is it possible to have a more flexible normalization? (Avoid duplicated normalizations)
```
@ philipc2 Can we avoid duplicate normalization here? For example if the nodes are already normalized, we will not normalize them.

For example, set an attributes in the grid `node_x`,`node_y`,`node_z` to indicate if they're already normalized or not.

philipc2 replied:
I believe that we handle this here

https://github.com/UXARRAY/uxarray/blob/a6aa6294975f189e8c36dd3d14da745526e53e06/uxarray/grid/grid.py#L1983

rajeeja commented:
ASV Benchmark failing?

philipc2 commented:
pre-commit.ci autofix

hongyuchen1030 commented:
@philipc2 

The benchmark still doesn't work out for this PR, can you take a look? Thanks

rajeeja reviewed (approved):
Approving, we can fix the benchmarking for forked repo's later.

philipc2 commented on uxarray/core/zonal.py:
```diff
@@ -18,23 +17,15 @@ def _compute_non_conservative_zonal_mean(uxda, latitudes, use_robust_weights=Fal
-
     bounds = uxgrid.bounds.values
+    faces_edges_cartesian = uxgrid.faces_edges_cartesian.values
 
     for i, lat in enumerate(latitudes):
         face_indices = uxda.uxgrid.get_faces_at_constant_latitude(lat)
 
         z = np.sin(np.deg2rad(lat))
 
-        faces_edge_nodes_xyz_candidate = faces_edge_nodes_xyz[face_indices, :, :, :]
+        faces_edge_nodes_xyz_candidate = faces_edges_cartesian[face_indices, :, :, :]
```
Wouldn't a better implementation here be to only construct the `faces_edges_cartesian` for the indices that we require, instead of needing to construct the entire array and then index?

For each line of constant latitude, typically only a small percentage of the total faces will be candidates.

hongyuchen1030 replied:
> Something like the following should work if we put it inside of the loop?
> 
> ```python
> _get_cartesian_face_edge_nodes(
>        uxgrid.face_node_connectivity.values[face_indices], # only candidate faces
>        len(face_indices), # number of candidate faces
>        uxgrid.n_max_face_nodes,
>        uxgrid.node_x.values,
>        uxgrid.node_y.values,
>        uxgrid.node_z.values,
>     )
> ```

Please refer to my comment below: the lalon bounds building will required this connectivity for the entire grid anyway

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1443,6 +1445,26 @@ def face_areas(self, value):
         assert isinstance(value, xr.DataArray)
         self._ds["face_areas"] = value
 
+    @property
+    def faces_edges_cartesian(self):
+        """Cartesian Coordinates for each Face Edge.
+
+        Dimensions ``(n_face, n_max_face_edges, 2, 3)``
+        """
+        if "faces_edges_cartesian" not in self._ds:
+            _populate_faces_edges_cartesian(self)
+        return self._ds["faces_edges_cartesian"]
+
+    @property
+    def faces_edges_spherical(self):
+        """Latitude Longitude Coordinates for each Face in radians.
+
+        Dimensions ``(n_face, n_max_face_edges, 2, 2)``
+        """
+        if "faces_edges_spherical" not in self._ds:
+            _populate_faces_edges_spherical(self)
+        return self._ds["faces_edges_spherical"]
+
```
I'm not sure if caching these throughout the entire lifetime of the `Grid` is something we want to do. 

For the 42 million element MPAS grid, each of these would be multiple GB in size. 

From a previous discussion with @erogluorhan 

https://github.com/UXARRAY/uxarray/pull/1130#discussion_r1924731908

In my other comment, we should instead construct only the desired faces for these variables in the zonal average calls, instead of storing the pre-computed array.

Since this variable simply manipulates existing connectivity and coordinate variables, we won't be losing any information or doing any expensive re-computations.

hongyuchen1030 replied:
Hi Philip, thanks for your input. Just to clarify a few points:

1. This property is constructed **only** when needed to populate the face_latlon bounds, **similar to how it was handled in the old design.**
2. Previously, **we populated these** for the whole grid when constructing the latlon bounds.
3. Also, we never permanently saved this property. It was created more than once—once for the face_latlon bounds and then partially re-populated for the zonal average.
4. With the new version, we **only** cache the populated face_edge_nodes property when required for the latlon bounds, which helps avoid duplicate calls later in the process.

So, if you take a closer look at the previous implementation, you'll notice **we didn't actually populate any extra data for the entire grid**. Hope this clears things up!

philipc2 replied:
Thanks for the detailed response!

> This property is constructed only when needed to populate the face_latlon bounds, similar to how it was handled in the old design.

That's consistent with how we've handled `Grid` variables. My concern is that both of these new variables are extremely large, and inherently are derived from existing variables that are already stored in our `Grid` (`face_node_connectivity` and either set of coordinates).

Caching something like the face bounds is practical because the time needed to generate it is the main bottle-neck in the initial zonal average calls. Additionally, it only contains `(n_face x 4)` total entries, compared to the `(n_face x n_max_face_nodes x 2 x (2/3))` for the `edge_nodes` here 


> This property is constructed only when needed to populate the face_latlon bounds, similar to how it was handled in the old design.

This makes sense, but my issue lies in whether we need to store the `face_edges` after constructing the bounds. As I mentioned, for our high-resolution grids these arrays require a significant amount of memory. 


> With the new version, we only cache the populated face_edge_nodes property when required for the latlon bounds, which helps avoid duplicate calls later in the process.

I do see the intention with removing the duplicate calls, but right now the bottle-neck with the zonal average is the initial bounds construction. 

For a `30km` grid (~600k hexagons), the initial execution time for the default `.zonal_average()` call is ~33 seconds on my machine. Subsequent calls are about 400ms.

The initial call is dominated by the bounds construction.

philipc2 replied:
One area of optimization could be to parallelize this Python loop with Numba.

https://github.com/hongyuchen1030/uxarray/blob/29862c2823abe49b33942ba36961d9891a21bd38/uxarray/core/zonal.py#L23-L49

The zonal average for each line of constant latitude can be computed independently.

hongyuchen1030 replied:
Thanks for the detailed response! I'm glad we're on the same page regarding bounds construction—it’s an inevitable step, and we need the full property for it.

If we look into the current implementation closely, as we loop from -90° to 90° for candidate faces, we're essentially rebuilding the entire face_edge_nodes connectivity. 

If the concern is mainly about the memory footprint for storing this grid property, **we can consider removing it after the zonal average call.**

philipc2 replied:
> I'm glad we're on the same page regarding bounds construction—it’s an inevitable step, and we need the full property for it.

Yep! Having the bounds cached is extremely important.

> If we look into the current implementation closely, as we loop from -90° to 90° for candidate faces, we're essentially rebuilding the entire face_edge_nodes connectivity.

I consider the `face_edge_nodes` to be one step past a `connectivity` variable. Our definition of a connectivity variable is an array of indices that point to some geometry (nodes, edges, or faces).

When we enter `_compute_non_conservative_zonal_mean()`, we compute the `face_edge_nodes` before the `latitude` loop, so we have access to the entire array for each line of constant latitude. Once we finish the zonal average call, we no longer need to store the `face_edge_nodes`

Here's a breakdown of the zonal average call for the example I provided above.

![image](https://github.com/user-attachments/assets/fd0e37ac-2bc3-4b3a-981c-567dcbb8073c)

I think there is definitely some further optimization that we could do with this helper. It's not currently written with Numba. 

I'd like to see how much we could optimize this helper before we consider caching computed value.

hongyuchen1030 replied:
Yes, exactly. This is why I believe we should store the connectivity information throughout the entire zonal average call—it doesn't contradict anything you've mentioned.

First, let's think thoroughly about the routine. Here's the detailed breakdown for the zonal average call:

1. **Before the zonal average call:** We build the latlon bounds, which requires the complete sets of faces_edges_cartesian and faces_edges_spherical for the entire grid.
2. **During the zonal weight calculation:** As we loop from -90° to 90° for candidate faces, we effectively need the full grid's connectivity information again. Caching the connectivity saves us from rebuilding it for each slice.
3. **After the zonal average call:** Once we're done, we remove the face_edge_nodes property since it's no longer needed.

Additionally, regarding parallelization: rather than focusing solely on paralleling each latitude slice, we should tackle the bottleneck—the latlon bound construction. Caching the connectivity also aids in parallelization, as each thread can read its slice of face_edge_nodes data directly without needing to rebuild it.

Also, I think it's important to consider the big picture. Instead of honing in on individual helper functions, we should think about the remapping operation as a whole. This integrated pipeline approach will give us a better framework for redesigning the process, rather than treating each function in isolation.

erogluorhan replied:
Sorry I didn't notice before that I was pinged here. My thoughts are:

I see the motive behind avoiding duplicate calls as much as we can. The memory consumption s also another limiting factor, so we should be looking for the sweet spot between performance vs. memory optimizations to provide the best user experience. 

That said, the discussion in https://github.com/UXARRAY/uxarray/pull/1130#discussion_r1924731908 is important I believe. You two, Hongyu and Philip, may know the details of this PR much better than me, but my first impressions is that `faces_edges_spherical ` is being generated from an existing (or once-generated) connectivity info, and its implementation seems vectorized (despite presence of some potentially expensive Numpy functions like unraveling in it), so I'd be more worried on the added memory consumption here than avoiding duplicate calls.

I am happy to look into more details of the code if needed though.

hongyuchen1030 commented:
Hi @philipc2 ,

Thanks again for the discussion—I'm glad we reached an agreement. I've added the latest commit, which removes the "faces_edges_cartesian" and "faces_edges_spherical" properties at the end of the zonal_mean call. This change should help save duplicate call time and prevent unnecessary memory expansion.

Please take a look when you have a moment. Once this is merged, I'll open another PR where we can integrate and carefully parallelize the latlonbound and zonal_avg calls.

Thanks!

hongyuchen1030 commented:
And I also thought carefully about optimizing the loop you mentioned:

> Something like the following should work if we put it inside of the loop?
> 
> ```python
> _get_cartesian_face_edge_nodes(
>        uxgrid.face_node_connectivity.values[face_indices], # only candidate faces
>        len(face_indices), # number of candidate faces
>        uxgrid.n_max_face_nodes,
>        uxgrid.node_x.values,
>        uxgrid.node_y.values,
>        uxgrid.node_z.values,
>     )
> ```


I know it sounds good to build the connectivity for each face slice inside the loop, and it would be very helpful if the user is querying just a few latitudes. However, consider the case of a batch of latitudes: for every latitude call, we would be querying the associated `uxgrid.node_x.values`, `uxgrid.node_y.values`, and `uxgrid.node_z.values` for the candidate faces, converting them into Cartesian face edge nodes, and then discarding them. Meanwhile, we already have this information available from the `latlon_bound` build.

This brings up another general design idea (which also applies to S2 geometry): we should separate the single call from the batch call. Our current version works best for a batch call for candidate faces—which is a common use case. In a future optimization, we can also consider optimizing the function for a single latitude call, perhaps even creating a dedicated function tailored for that scenario.

Let me know your thoughts.

philipc2 commented:
> @philipc2 And I also thought carefully about optimizing the loop you mentioned:
> 
> > Something like the following should work if we put it inside of the loop?
> > ```python
> > _get_cartesian_face_edge_nodes(
> >        uxgrid.face_node_connectivity.values[face_indices], # only candidate faces
> >        len(face_indices), # number of candidate faces
> >        uxgrid.n_max_face_nodes,
> >        uxgrid.node_x.values,
> >        uxgrid.node_y.values,
> >        uxgrid.node_z.values,
> >     )
> > ```
> 
> I know it sounds good to build the connectivity for each face slice inside the loop, and it would be very helpful if the user is querying just a few latitudes. However, consider the case of a batch of latitudes: for every latitude call, we would be querying the associated `uxgrid.node_x.values`, `uxgrid.node_y.values`, and `uxgrid.node_z.values` for the candidate faces, converting them into Cartesian face edge nodes, and then discarding them. Meanwhile, we already have this information available from the `latlon_bound` build.
> 

This is a good point. Keeping the call outside of the loop should be good for our current application. 

> This brings up another general design idea (which also applies to S2 geometry): we should separate the single call from the batch call. Our current version works best for a batch call for candidate faces—which is a common use case. In a future optimization, we can also consider optimizing the function for a single latitude call, perhaps even creating a dedicated function tailored for that scenario.

I like that idea.

philipc2 commented on uxarray/grid/grid.py:
```diff
@@ -1443,6 +1445,44 @@ def face_areas(self, value):
         self._ds["face_areas"] = value
 
+    @property
+    def faces_edges_cartesian(self):
+        """Cartesian Coordinates for each Face Edge.
+
+        Dimensions ``(n_face, n_max_face_edges, 2, 3)``
+        """
+        if "faces_edges_cartesian" not in self._ds:
+            _populate_faces_edges_cartesian(self)
+        return self._ds["faces_edges_cartesian"]
+
+    @faces_edges_cartesian.setter
+    def faces_edges_cartesian(self, value):
+        self._ds["faces_edges_cartesian"] = value
+
+    @faces_edges_cartesian.deleter
+    def faces_edges_cartesian(self):
+        if "faces_edges_cartesian" in self._ds:
+            del self._ds["faces_edges_cartesian"]
+
+    @property
+    def faces_edges_spherical(self):
+        """Latitude Longitude Coordinates for each Face in radians.
+
+        Dimensions ``(n_face, n_max_face_edges, 2, 2)``
+        """
+        if "faces_edges_spherical" not in self._ds:
+            _populate_faces_edges_spherical(self)
+        return self._ds["faces_edges_spherical"]
+
+    @faces_edges_spherical.setter
+    def faces_edges_spherical(self, value):
+        self._ds["faces_edges_spherical"] = value
+
+    @faces_edges_spherical.deleter
+    def faces_edges_spherical(self):
+        if "faces_edges_spherical" in self._ds:
+            del self._ds["faces_edges_spherical"]
+
```
This is a step in the right direction, although I don't think we should store the `face_edges` in `self._ds` due to the memory concerns.

I'm happy to keep the two `faces_edges_spherical` and `faces_edges_cartesian` as part of the API, but instead they construct the variables by calling `_get_cartesian_faces_edge_nodes`.

Inside of `zonal.py`, calling this attribute within the scope of the function will automatically manage the memory associated with it once we exit the function.

hongyuchen1030 replied:
Keeping `face_edges` in zonal was my first thought as well. However, I then realized that l`atlonbound `also needs to use this property. That means we would have to modify the `latlonbound` API to accept incoming face_edges, which involves more API changes than I initially expected—and I’m wary of making such significant modifications to the existing API and functionality.

Do you have any suggestions on where we might store this information to address the memory concerns without causing a large API overhaul?

Thanks!

philipc2 replied:
We can keep the two `@property` implementations, and have them return the output of `_get_cartesian_face_edge_nodes()` and `_get_lonlat_rad_face_edge_nodes()` and replace the following two lines with calls to `Grid.face_edges_cartesian` and `Grid.face_edges_spherical`

https://github.com/UXARRAY/uxarray/blob/a6aa6294975f189e8c36dd3d14da745526e53e06/uxarray/grid/geometry.py#L1412-L1428

philipc2 replied:
@hongyuchen1030 

Were you able to make this change?

hongyuchen1030 replied:
I am unable to find such a leftover design in this PR. I have also pinned you to the corresponding `populate_bounds` section because the whole point of the PR was to reduce such duplicate calls. I wonder which commit this file belongs to?

philipc2 commented:
@hongyuchen1030 

Here's an optimized version of `get_cartesian_face_edge_nodes` using Numba. Here we re-use the `face_edge_connectivity` and `edge_node_connectivity` that can be computed internally and re-used.

This runs in about `90ms` for me, down from the `~500ms` we currently have. 



```Python
@njit(cache=True, parallel=True)
def get_cartesian_face_edge_nodes(face_edge_connectivity, edge_node_connectivity,
                                  node_x, node_y, node_z,
                                  n_edges_per_face, n_face, n_max_face_edges):

    # Allocate result array
    # I used np.nan here instead of INT_FILL_VALUE to be consistent with the double type (INT_FILL_VALUE is for our int64 arrays)
    face_edges = np.full((n_face, n_max_face_edges, 2, 3), np.nan, dtype=np.float64)

    # Construct edge_nodes for each face in parallel
    for face_i in prange(n_face):
        n_edges = n_edges_per_face[face_i]
        edge_indices = face_edge_connectivity[face_i, :n_edges]
        node_indices = edge_node_connectivity[edge_indices]  

        first_nodes = node_indices[:, 0]
        second_nodes = node_indices[:, 1]

        # Assign x coordinates
        face_edges[face_i, :n_edges, 0, 0] = node_x[first_nodes]
        face_edges[face_i, :n_edges, 1, 0] = node_x[second_nodes]

        # Assign y coordinates
        face_edges[face_i, :n_edges, 0, 1] = node_y[first_nodes]
        face_edges[face_i, :n_edges, 1, 1] = node_y[second_nodes]

        # Assign z coordinates
        face_edges[face_i, :n_edges, 0, 2] = node_z[first_nodes]
        face_edges[face_i, :n_edges, 1, 2] = node_z[second_nodes]

    return face_edges

```

The same can be applied to the spherical one, with changes only needed to the assignment of coordinates.

hongyuchen1030 commented:
> ```python
> face_node_connectivity
> ```

Thanks for the information! This is actually what I had in mind for `get_cartesian_face_edge_nodes` using `face_edge_connectivity` and `edge_node_connectivity`. However, I recall that we had a lengthy discussion about why we should separate the steps—first populating `face_node_connectivity` (using  `face_edge_connectivity` and `edge_node_connectivity`) and storing it in `_ds`, and then populating face_edge_connectivity when needed.

Do you still recall the rationale behind that decision? I'm trying to avoid duplicate iterations as much as possible.

Thanks!

philipc2 commented:
> > ```python
> > face_node_connectivity
> > ```
> 
> Thanks for the information! This is actually what I had in mind for `get_cartesian_face_edge_nodes` using `face_edge_connectivity` and `edge_node_connectivity`. However, I recall that we had a lengthy discussion about why we should separate the steps—first populating `face_node_connectivity` (using `face_edge_connectivity` and `edge_node_connectivity`) and storing it in `_ds`, and then populating face_edge_connectivity when needed.
> 
> Do you still recall the rationale behind that decision? I'm trying to avoid duplicate iterations as much as possible.
> 
> Thanks!

Most of the time, the `face_node_connectivity` is already present and we do not need to construct it. The other two connectivity variables are automatically populated when we call their respective attribute.

It makes sense to me to re-use those connectivity variables instead of doing the extra operations within the function call, which each time essentially "derives" the same information already present in the connectivity.

hongyuchen1030 commented:
Hi Kelly,

Thank you for the detailed explanation and your help. I appreciate the
guidance regarding the pairing requirements.

I will reach out again next week about both semi-private and private lesson
options. If a suitable pairing for the semi-private lesson isn’t available,
I’m happy to continue with a private lesson.

Thanks again for your assistance.

Best regards,

Victoria

On Tue, Apr 1, 2025 at 12:50 PM Philip Chmielowiec ***@***.***>
wrote:

> face_node_connectivity
>
> Thanks for the information! This is actually what I had in mind for
> get_cartesian_face_edge_nodes using face_edge_connectivity and
> edge_node_connectivity. However, I recall that we had a lengthy
> discussion about why we should separate the steps—first populating
> face_node_connectivity (using face_edge_connectivity and
> edge_node_connectivity) and storing it in _ds, and then populating
> face_edge_connectivity when needed.
>
> Do you still recall the rationale behind that decision? I'm trying to
> avoid duplicate iterations as much as possible.
>
> Thanks!
>
> Most of the time, the face_node_connectivity is already present and we do
> not need to construct it. The other two connectivity variables are
> automatically populated when we call their respective attribute.
>
> It makes sense to me to re-use those connectivity variables instead of
> doing the extra operations within the function call, which each time
> essentially "derives" the same information already present in the
> connectivity.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UXARRAY/uxarray/pull/1180#issuecomment-2770527388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASK37KQUWKVAWYBJTZFZRSL2XLUYNAVCNFSM6AAAAABY2L2ISWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZQGUZDOMZYHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: philipc2]*philipc2* left a comment (UXARRAY/uxarray#1180)
> <https://github.com/UXARRAY/uxarray/pull/1180#issuecomment-2770527388>
>
> face_node_connectivity
>
> Thanks for the information! This is actually what I had in mind for
> get_cartesian_face_edge_nodes using face_edge_connectivity and
> edge_node_connectivity. However, I recall that we had a lengthy
> discussion about why we should separate the steps—first populating
> face_node_connectivity (using face_edge_connectivity and
> edge_node_connectivity) and storing it in _ds, and then populating
> face_edge_connectivity when needed.
>
> Do you still recall the rationale behind that decision? I'm trying to
> avoid duplicate iterations as much as possible.
>
> Thanks!
>
> Most of the time, the face_node_connectivity is already present and we do
> not need to construct it. The other two connectivity variables are
> automatically populated when we call their respective attribute.
>
> It makes sense to me to re-use those connectivity variables instead of
> doing the extra operations within the function call, which each time
> essentially "derives" the same information already present in the
> connectivity.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/UXARRAY/uxarray/pull/1180#issuecomment-2770527388>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASK37KQUWKVAWYBJTZFZRSL2XLUYNAVCNFSM6AAAAABY2L2ISWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZQGUZDOMZYHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

hongyuchen1030 commented:
> @hongyuchen1030
> 
> Here's an optimized version of `get_cartesian_face_edge_nodes` using Numba. Here we re-use the `face_edge_connectivity` and `edge_node_connectivity` that can be computed internally and re-used.
> 
> This runs in about `90ms` for me, down from the `~500ms` we currently have.
> 
> ```python
> @njit(cache=True, parallel=True)
> def get_cartesian_face_edge_nodes(face_edge_connectivity, edge_node_connectivity,
>                                   node_x, node_y, node_z,
>                                   n_edges_per_face, n_face, n_max_face_edges):
> 
>     # Allocate result array
>     # I used np.nan here instead of INT_FILL_VALUE to be consistent with the double type (INT_FILL_VALUE is for our int64 arrays)
>     face_edges = np.full((n_face, n_max_face_edges, 2, 3), np.nan, dtype=np.float64)
> 
>     # Construct edge_nodes for each face in parallel
>     for face_i in prange(n_face):
>         n_edges = n_edges_per_face[face_i]
>         edge_indices = face_edge_connectivity[face_i, :n_edges]
>         node_indices = edge_node_connectivity[edge_indices]  
> 
>         first_nodes = node_indices[:, 0]
>         second_nodes = node_indices[:, 1]
> 
>         # Assign x coordinates
>         face_edges[face_i, :n_edges, 0, 0] = node_x[first_nodes]
>         face_edges[face_i, :n_edges, 1, 0] = node_x[second_nodes]
> 
>         # Assign y coordinates
>         face_edges[face_i, :n_edges, 0, 1] = node_y[first_nodes]
>         face_edges[face_i, :n_edges, 1, 1] = node_y[second_nodes]
> 
>         # Assign z coordinates
>         face_edges[face_i, :n_edges, 0, 2] = node_z[first_nodes]
>         face_edges[face_i, :n_edges, 1, 2] = node_z[second_nodes]
> 
>     return face_edges
> ```
> 
> The same can be applied to the spherical one, with changes only needed to the assignment of coordinates.

Hi @philipc2 

Have you compared the new version's results with the old one? I set up the following test and found that the results differ:
```python
def test_cartesian_test():
    grid = ux.open_grid(gridfile_CSne8)
    new_faces_edges_cartesian = get_cartesian_face_edge_nodes(
        grid.face_edge_connectivity.values,
        grid.edge_node_connectivity.values,
        grid.node_x.values,
        grid.node_y.values,
        grid.node_z.values,
        grid.n_edges_per_face,  # or grid.n_face, depending on what you intended
        grid.n_max_face_edges
    )

    old_faces_edges_cartesian = _get_cartesian_faces_edge_nodes(
        grid.face_node_connectivity.values, grid.n_face,
        grid.n_max_face_edges, grid.node_x.values,
        grid.node_y.values, grid.node_z.values
    )

    nt.assert_array_equal(new_faces_edges_cartesian, old_faces_edges_cartesian)
```
The error messages are:

```python
args = (<built-in function eq>, array([[[[ 0.63933827, -0.42719217, -0.63933827],
         [ 0.57735027, -0.57735027, -0.5773... 0.63933827]],

        [[-0.63933827,  0.42719217,  0.63933827],
         [-0.48565274,  0.48565274,  0.72683068]]]]))
kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 4516 / 9216 (49%)
E           Max absolute difference among violations: 0.19509032
E           Max relative difference among violations: 1.60290531e+16
E            ACTUAL: array([[[[ 0.639338, -0.427192, -0.639338],
E                    [ 0.57735 , -0.57735 , -0.57735 ]],
E           ...
E            DESIRED: array([[[[ 0.57735 , -0.57735 , -0.57735 ],
E                    [ 0.639338, -0.427192, -0.639338]],
E           ...

```

philipc2 commented:
> > @hongyuchen1030
> > Here's an optimized version of `get_cartesian_face_edge_nodes` using Numba. Here we re-use the `face_edge_connectivity` and `edge_node_connectivity` that can be computed internally and re-used.
> > This runs in about `90ms` for me, down from the `~500ms` we currently have.
> > ```python
> > @njit(cache=True, parallel=True)
> > def get_cartesian_face_edge_nodes(face_edge_connectivity, edge_node_connectivity,
> >                                   node_x, node_y, node_z,
> >                                   n_edges_per_face, n_face, n_max_face_edges):
> > 
> >     # Allocate result array
> >     # I used np.nan here instead of INT_FILL_VALUE to be consistent with the double type (INT_FILL_VALUE is for our int64 arrays)
> >     face_edges = np.full((n_face, n_max_face_edges, 2, 3), np.nan, dtype=np.float64)
> > 
> >     # Construct edge_nodes for each face in parallel
> >     for face_i in prange(n_face):
> >         n_edges = n_edges_per_face[face_i]
> >         edge_indices = face_edge_connectivity[face_i, :n_edges]
> >         node_indices = edge_node_connectivity[edge_indices]  
> > 
> >         first_nodes = node_indices[:, 0]
> >         second_nodes = node_indices[:, 1]
> > 
> >         # Assign x coordinates
> >         face_edges[face_i, :n_edges, 0, 0] = node_x[first_nodes]
> >         face_edges[face_i, :n_edges, 1, 0] = node_x[second_nodes]
> > 
> >         # Assign y coordinates
> >         face_edges[face_i, :n_edges, 0, 1] = node_y[first_nodes]
> >         face_edges[face_i, :n_edges, 1, 1] = node_y[second_nodes]
> > 
> >         # Assign z coordinates
> >         face_edges[face_i, :n_edges, 0, 2] = node_z[first_nodes]
> >         face_edges[face_i, :n_edges, 1, 2] = node_z[second_nodes]
> > 
> >     return face_edges
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The same can be applied to the spherical one, with changes only needed to the assignment of coordinates.
> 
> Hi @philipc2
> 
> Have you compared the new version's results with the old one? I set up the following test and found that the results differ:
> 
> ```python
> def test_cartesian_test():
>     grid = ux.open_grid(gridfile_CSne8)
>     new_faces_edges_cartesian = get_cartesian_face_edge_nodes(
>         grid.face_edge_connectivity.values,
>         grid.edge_node_connectivity.values,
>         grid.node_x.values,
>         grid.node_y.values,
>         grid.node_z.values,
>         grid.n_edges_per_face,  # or grid.n_face, depending on what you intended
>         grid.n_max_face_edges
>     )
> 
>     old_faces_edges_cartesian = _get_cartesian_faces_edge_nodes(
>         grid.face_node_connectivity.values, grid.n_face,
>         grid.n_max_face_edges, grid.node_x.values,
>         grid.node_y.values, grid.node_z.values
>     )
> 
>     nt.assert_array_equal(new_faces_edges_cartesian, old_faces_edges_cartesian)
> ```
> 
> The error messages are:
> 
> ```python
> args = (<built-in function eq>, array([[[[ 0.63933827, -0.42719217, -0.63933827],
>          [ 0.57735027, -0.57735027, -0.5773... 0.63933827]],
> 
>         [[-0.63933827,  0.42719217,  0.63933827],
>          [-0.48565274,  0.48565274,  0.72683068]]]]))
> kwds = {'err_msg': '', 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
> 
>     @wraps(func)
>     def inner(*args, **kwds):
>         with self._recreate_cm():
> >           return func(*args, **kwds)
> E           AssertionError: 
> E           Arrays are not equal
> E           
> E           Mismatched elements: 4516 / 9216 (49%)
> E           Max absolute difference among violations: 0.19509032
> E           Max relative difference among violations: 1.60290531e+16
> E            ACTUAL: array([[[[ 0.639338, -0.427192, -0.639338],
> E                    [ 0.57735 , -0.57735 , -0.57735 ]],
> E           ...
> E            DESIRED: array([[[[ 0.57735 , -0.57735 , -0.57735 ],
> E                    [ 0.639338, -0.427192, -0.639338]],
> E           ...
> ```

Looks like there is some small differences. We can move this optimizing into the work I'm putting together in #1195 and keep this PR smaller in scope.

hongyuchen1030 commented:
> Looks like there is some small differences. We can move this optimizing into the work I'm putting together in #1195 and keep this PR smaller in scope.

@philipc2 
Sounds good, Can you approve this PR so it can be merged? Thanks

philipc2 commented:
> > Looks like there is some small differences. We can move this optimizing into the work I'm putting together in #1195 and keep this PR smaller in scope.
> 
> @philipc2 Sounds good, Can you approve this PR so it can be merged? Thanks

Yep! Wrapping up a final review.

hongyuchen1030 commented on uxarray/grid/geometry.py:
```diff
@@ -1407,25 +1466,14 @@ def _populate_bounds(
-    )
+    # If the face_edges_cartesian and face_edges_lonlat_rad didn't exist,
+    # The following call will automatically populate them
+    faces_edges_cartesian = grid.faces_edges_cartesian.values
```
@philipc2 

So in this PR, it's already replaced with the call to `grid.faces_edges_cartesian.values` and `grid.faces_edges_spherical.values`

philipc2 reviewed (changes requested):
Just to clarify a couple things.

I would prefer not to handle the deletion of the allocated arrays through the `Grid` attributes. 

If we want to keep the attributes (i.e. `Grid.faces_edges_cartesian`), these should not store anything as part of the `Grid._ds`. We would instead use them for convenience within our calls (i.e. what you've already done inside of zonal average).

This loops back to the discussion I had previously with @erogluorhan . Pinging him here if he has anything to add.

I do see the benefit of caching and don't mean to invalidate the optimizations in this PR, however the current issue that is still making the usage of some of our functionality impractical for the highest resolution grids is still the execution time of `Grid.bounds`

I hope this makes sense.

erogluorhan commented on uxarray/core/zonal.py:
```diff
@@ -18,23 +17,15 @@ def _compute_non_conservative_zonal_mean(uxda, latitudes, use_robust_weights=Fal
-    )
-
     bounds = uxgrid.bounds.values
+    faces_edges_cartesian = uxgrid.faces_edges_cartesian.values
```
I've seen a lot of these `.values` calls in this PR. I see that conversion to Numpy arrays is required for Numba executions, but I want to emphasize a couple things here:

- Are all of the `.values` conversions in this code required for the Numba injections here. Could any of them be avoided?
   - Asking this because Xarray-to-Numpy conversions, i.e. `.values` calls, are dangerous when we have chunked Xarray inputs, which may otherwise not fit into the memory. These `.values` may break the executions in such cases since they will attempt to convert the whole data into Numpy arrays, which couldn't fit into memory.
- Even for those that would be required for Numba calls, we will still need to find a robust longer-term solution, which will probably be a combination of `map_blocks()` or `apply_ufunc()` with `.data` calls where chunked Dask arrays would be efficiently processed through one of these methods.

hongyuchen1030 replied:
Hi @erogluorhan ,

These `.values` calls aren’t something new introduced in this PR— If you read the current implementation, you will notice that they follow the existing implementation in the `bounds` function. We’re simply maintaining the same approach here. If you have suggestions for a better long-term solution to handle chunked inputs more robustly (e.g., through `map_blocks()` or `apply_ufunc()` with `.data`), We can open a new PR to address those overall design improvements.

Thanks for bringing this up!

erogluorhan replied:
Thanks @hongyuchen1030 ! Just to clarify, yes, I can see those `.values` in the older code, but I just wanted to remind how keeping those `.values` calls in the code would be dangerous for larger-than-memory data. 

So, even though we've had them in the code so far, we will need to avoid them (or get rid of the existing ones) as much as we can in our future code to fully optimize UXarray for memory usage and parallelization.  

If you want to keep this PR's scope limited to what you've done already though, I am just fine with it. Things I am pointing out can be handled in separate work.

See the [Consistently use Xarray for Grid Reader, Improved Dask Support(https://github.com/UXARRAY/uxarray/pull/1092) PR, which was one of the most recent optimization work we had, you haven't seen yet. This work didn't involve any Numba though, so it was relatively easier to handle. 

To handle Numba and this together, we'll need to look into mechanisms I mentioned in the first comment.

hongyuchen1030 replied:
Thanks for the information, @erogluorhan .

I agree that addressing the extensive use of `.values` throughout UXarray is a significant task that goes beyond the scope of this PR—it essentially calls for an overall redesign of the codebase to optimize memory usage and parallelization.

 I believe the points you've raised should be tracked in a new PR or issue dedicated to this redesign. That way, we can ensure these considerations are not lost and are properly addressed in a future effort.

For now, I'll keep the changes in this PR as-is.

Thanks again for your insights!

erogluorhan replied:
Sounds good to me; thanks @hongyuchen1030 !

philipc2 commented:
I apologize if I've caused any confusion.


> I understand that you’d prefer not to keep these variables in Grid._ds, but you previously mentioned that it’s okay to have them as properties there. So I'm not entirely clear on where you’d like to store them instead.

Our properties don't need to return cached variables only. We could simply wrap the building functions for `face_edges_cartesian` and `face_edges_spherical` and use those in our implementations. This is operationally the same, but reduces the overall length of our code.

I do see the benefit in caching the variables, especially considering the breakoutof the zonal average call above. However, before we chose to cache and re-use this values, I believe that optimizing the helper function should be the first step.

Since constructing the `face_edges` only involves the manipulation of connectivity variables and indexing existing coordinates, I'd prefer not to use up extra memory throughout the life-span of the Grid.

The current issue with how you've implemented things is that it assumes the user is going to immediately call `.zonal_average()` after constructing the bounds, which isn't always the case. 

Also, the performance of our Zonal Average functionality is at a good spot right now. The current bottle neck still lies in the initial bounds construction, which has come a long way since our initial implementation, but could still use extra optimizations. The change here would not lead to a noticeable change in perfromance.


| Resolution | Nodes      | Faces      | Edges       | Grid Load Time (s) | Bounds Construction Time (s) | Total Time (s) |
|------------|------------|------------|-------------|-------------------:|----------------------------:|---------------:|
| 30km       | 1,310,720  | 655,362    | 1,966,080   | 1.923              | 2.541                      | 4.464          |
| 15km       | 5,242,880  | 2,621,442  | 7,864,320   | 7.489              | 10.191                     | 17.68          |
| 7.5km      | 20,971,520 | 10,485,762 | 31,457,280  | 29.228             | 37.866                     | 67.095         |
| 3.75km     | 83,886,080 | 41,943,042 | 125,829,120 | 113.028            | 155.486                    | 268.514        |

To summarize:
- We should investigate further optimizations into constructing the `face_edge_` variables before considering caching
  - If we reach a wall here, we could consider adding a configuration parameter that the user can select if they would like to consider caching. 
- We should continue to optimize the Bounds construction, since this is the most noticeable slowdown (even though it is a one-time execution) 

I am happy to have a discussion about this during this weeks UXarray meeting.

hongyuchen1030 commented:
> That is understandable, however having the bounds is a valuable addition to the user API, as it has applications to visualization and subsetting/cross-section functionality that we have implemented.

I can proceed to revoke the changes. 

Just some comments on this:
>That is understandable, however having the bounds is a valuable addition to the user API, as it has applications to visualization and subsetting/cross-section functionality that we have implemented.

This still means the latlonbound is served as an intermediate process. And the right way to do is to optimize the subsetting/cross-section functionality as a whole, and even provide a tailored latlonbounds building for these functionalities in the future

philipc2 commented:
@hongyuchen1030 

I am putting together optimizations to the constructing of `Grid.bounds` in #1205 to compute the face edge nodes for each face as they are needed, instead of pre-computing the entire array.

Once that is merged, we should considering doing something similar for the zonal face weights. We could tackle that as part of this PR.

erogluorhan commented:
@hongyuchen1030 given that @philipc2 's improvements have gone into release, should we close this one and its corresponding issue #1181?

erogluorhan commented:
Confirmed with @hongyuchen1030 that this can be closed.
